### PR TITLE
Fix incorrect usage of tracing span in async functions.

### PR DESCRIPTION
**Pull request by robin-near**
See https://docs.rs/tracing/latest/tracing/struct.Span.html#method.enter

It is incorrect to do this:

```
async fn f() {
  let _ = tracing::span!(...).entered();
  ...
}
```

because when the function yields, the tracing span remains entered. This causes incorrect span nesting, and we have observed a stack overflow as a result, when the system is under load.

### CODE REVIEW

**Comment by robin-near** [comment]
@nagisa Any idea how to make this lintable? It's pretty subtle and I don't imagine this is very common knowledge. On the other hand, it's not such a huge deal because we don't use a lot async functions outside of networking code. Still, one day someone is going to do this again.

**Comment by akhi3030** [comment]
I wonder if it is possible to check statically if you are inside an async function.  If it possible to do that, then we could introduce that check around the call to span.  Would something like that work?

**Comment by nagisa** [comment]
I don't think there are easy ways to prevent `tracing::Span`s from existing across yield points without (breaking) changes to `tracing` itself.

One way would be to have `Span` not be `Send+Sync`, which with our threaded use would make the futures `!Send !Sync` as well.

More reasonably, this wouldn't have occurred if we chosen to use the [`#[instrument]`](https://docs.rs/tracing/latest/tracing/attr.instrument.html) macro instead of `span!` everywhere. It transparently does the correct thing based on the function type, and any use of other stuff would appear pretty out-of-place, which would suggest reviewers to pay a closer attention.

I would suggest doing exactly that in this instance instead of manually unrolling `async {}`s.

---

> It's pretty subtle and I don't imagine this is very common knowledge.

The warning to that effect is alluded to within the [first page of the `tracing` crate]( https://docs.rs/tracing/latest/tracing/index.html#spans). It is pretty hard for the project to do much else to draw awareness to the issue. I can only say that it is *not* uncommon to see people asking about invalid traces, but it is also not particularly common in absolute terms, because in most cases people just use `#[instrument]` and in other cases async use in a project is so pervasive, that it is hardly possible to not be aware of the gotcha.

All that said this is my first time seeing this sort of misuse becoming a stack overflow. I would consider sending a bug report to the upstream project(s) with a minimal reproducer they could experiment with. If you have the time for it, of course.

File: chain/network/src/peer/peer_actor.rs [hunk 1]
```diff
@@ -937,95 +937,105 @@ impl PeerActor {
...
+                }
+            })
+        };
+        // DO NOT turn this into a scoped .entered() span, because that is incompatible
+        // with async fns.
+        result
+            .instrument(tracing::trace_span!(
+                target: "network",
+                "receive_routed_message",
+                "type" = body_type,
```
Comment (nagisa) [inline chain/network/src/peer/peer_actor.rs:1036]: In this case `#[instrument]` application would need to use an initially empty field and then record it later as demonstrated in the [last example here](https://docs.rs/tracing/latest/tracing/attr.instrument.html#examples-1).

**Review by akhi3030 — APPROVED** [review]
Approving to unblock.

**Comment by robin-near** [comment]
I'm merging this to unblock the stack overflow issue, but not improving upon it as I don't have the bandwidth right now. Thanks @nagisa for your thoughts!

**Pull request merged**